submission/races.go

```diff
@@ -0,0 +1,299 @@
+// Re-calculates group-completion and cancels any running but not-awaited-anymore
+/// Log-requests.
+func (sub *safeSubmissionState) handleResult(logURL string, sct *ct.SignedCertificateTimestamp, err error) {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	if sct == nil {
+		sub.results[logURL] = &submissionResult{sct: sct, err: err}
+		return
+	}
+	// If at least one group needs that SCT, result is set. Otherwise ignored.
+	for groupName := range sub.logToGroups[logURL] {
+		if sub.groupNeeds[groupName] > 0 {
+			sub.results[logURL] = &submissionResult{sct: sct, err: err}
+		}
+		sub.groupNeeds[groupName]--
+	}
+
+	// Cancel any pending Log-requests for which there's no more awaiting Log-group.
+	for logURL, groupSet := range sub.logToGroups {
+		isAwaitedByGroup := false
+		for g := range groupSet {
+			if sub.groupNeeds[g] > 0 {
+				isAwaitedByGroup = true
+				break
+			}
+		}
+		if !isAwaitedByGroup && sub.cancels[logURL] != nil {
+			sub.cancels[logURL]()
+			delete(sub.cancels, logURL)
+		}
+	}
+}
+
+// returns whether the group has stopped waiting for more SCTs.
+func (sub *safeSubmissionState) groupComplete(groupName string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	needs, ok := sub.groupNeeds[groupName]
+	if !ok {
+		return true
+	}
+	return needs <= 0
+}
+
+func (sub *safeSubmissionState) collectSCTs() []*AssignedSCT {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	scts := []*AssignedSCT{}
+	for logURL, r := range sub.results {
+		if r != nil && r.sct != nil {
+			scts = append(scts, &AssignedSCT{LogURL: logURL, SCT: r.sct})
+		}
+	}
+	return scts
+}
+
+// submitInterval calculates duration for consequent call.
+// For first parallelStart calls duration is 0, while every next one gets
+// additional dur interval.
+func submitInterval(idx int, parallelStart int, dur time.Duration) time.Duration {
+	if idx < parallelStart {
+		return time.Duration(0)
+	}
+	return time.Duration(idx+1-parallelStart) * dur
+}
+
+// Submitter is interface wrapping Log-request-response cycle and any processing.
+type Submitter interface {
+	SubmitToLog(ctx context.Context, logURL string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error)
+}
+
+// groupRace shuffles logs within the group, submits avoiding duplicate-requests
+// and collects responses.
+func groupRace(ctx context.Context, chain []ct.ASN1Cert, group *ctpolicy.LogGroupInfo, parallelStart int,
+	state *safeSubmissionState, submitter Submitter) groupState {
+	groupURLs := make([]string, 0, len(group.LogURLs))
+	for logURL := range group.LogURLs {
+		groupURLs = append(groupURLs, logURL)
+	}
+
+	type count struct{}
+	counter := make(chan count, len(groupURLs))
+
+	// Randomize the order in which we send requests to the logs in a group
+	// so we maximize the distribution of logs we get SCTs from.
+	for i, urlNum := range rand.Perm(len(groupURLs)) {
+		subCtx, cancel := context.WithCancel(ctx)
+		go func(i int, logURL string) {
```

Comment:
I think it would have been easier to ingest if the parallel and sequential part of this algorithm were separate, and the delay mechanism was implemented outside of the goroutine. Starting 100 goroutines (if the group has that many) just to have 3-5 SCTs, for example, is not well justified under high load.

Also, can the delay mechanism be implemented using `time.Ticker`?

Reply:
Another thing: _I think_ that a reasonable assumption to make is that the first `parallelStart` calls will succeed in most cases, so having a quick path for this case is probably wanted.

Reply:
-- in current reality, it's impossible to get 100 goroutines as we don't get that many Logs. Moreover, here it's unlikely we have all of them set up as input. (Input is already pre-filtered)
-- the delay mechanism could be implemented as time.Ticker. But what's the advantage?

-- What do you mean by quick path? Currently sub-contexts are cancelled when enough results are returned.

Reply:
It's worth considering but but I think we should defer that sort of change.

I would also like to see the parallel parts split out more as I think it will be easier to understand and maintain. Or perhaps a simple "straw man" implementation at the moment that can be replaced later on.

Reply:
If this code is structured with explicit separation between the first multiple parallel calls, and the subsequent sequential calls, then fast path will come naturally. Pseudocode:

```golang
cctx, cancel := context.WithCancel(ctx)
for i := 0; i < parallel && i < len(group); i++ {
  go tryLog(cctx, i) // Might call cancel() when done.
}
ticker := time.NewTicker(delay)
for i := parallel; i < len(group); i++ {
  select {
  case <-ticker:
    go tryLog(i) // Might call cancel() when done.
  case <-cctx.Done():
    i = len(group)
  }
}
```
Also, some sort of `sync.WaitGroup` could potentially replace `counter{}` channel.

Reply:
I agree that this can be a TODO, with a follow-up PR focused on just that.